### PR TITLE
[WIP] Support single precision with MKL

### PR DESCRIPTION
**Pull request by niketanpansare**
- This PR contains the shared libraries for native BLAS functions compiled against MKL on 64-bit Linux (x86).
- It also contains a bugfix for the cmake setup.
- But the results for MKL single precision conv2d are incorrect. This can be tested with a LeNet script.

Since, I won't be able to work on that issue until Feb 16th (unless I find some free cycles in between or someone plans to take it over), I am marking this PR as WIP for time-being.

@mboehm7 @nakul02 @j143

### CODE REVIEW

**Comment by mboehm7** [comment]
Could it be that we need to memset(0) the output for MKL due to the reused direct float buffers? For the openblas conv2d realizations this was not necessary because the dense matrix multiply overwrites the output.

**Comment by mboehm7** [comment]
ok, let's do the following - I'll merge your change in to fix the nnz and cmake issues and I'll take care of the open issue.

**Comment by niketanpansare** [comment]
Interesting. We definitely need to fix cmake setup. For now, let’s use command line compilation for openblas 👍

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2269/

**Comment by mboehm7** [comment]
After creating myself an MKL setup, I gave it a try and interesting I get the same runtime issue when compiling with cmake. With a manual compilation it runs fine though - on lenet I get the following output (which looks fine to me):
`
18/02/02 21:48:23 INFO utils.NativeHelper: Using native blas: mkl
18/02/02 21:48:23 INFO mapred.FileInputFormat: Total input paths to process : 1
18/02/02 21:48:23 INFO mapred.FileInputFormat: Total input paths to process : 1
Iter:100, training loss:0.15647351276091165, training accuracy:95.3125
Iter:200, training loss:0.1730006931655677, training accuracy:92.1875
Iter:300, training loss:0.03702024917562899, training accuracy:98.4375
Iter:400, training loss:0.08554660386747788, training accuracy:98.4375
Iter:500, training loss:0.08748427048152668, training accuracy:95.3125
Iter:500, validation loss:317.60308121196874, validation accuracy:96.17300724637681
Iter:600, training loss:0.012787412034915456, training accuracy:100.0
Iter:700, training loss:0.031294041319592326, training accuracy:98.4375
Iter:800, training loss:0.043600472186685495, training accuracy:98.4375
Iter:900, training loss:0.01784090034780082, training accuracy:100.0
Iter:1000, training loss:0.00839880056975591, training accuracy:100.0
`

**Comment by niketanpansare** [comment]
It's surprising. Are you using single precision or double precision ?
